xds/internal/xdsclient/e2e_test/cds_watchers_test.go

```diff
@@ -0,0 +1,869 @@
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+	}
+
+	// Verify that the expected error is propagated to the watcher.
+	u, err := updateCh.Receive(ctx)
+	if err != nil {
+		t.Fatalf("timeout when waiting for a cluster resource from the management server: %v", err)
+	}
+	gotErr := u.(xdsresource.ClusterUpdateErrTuple).Err
+	if gotErr == nil || !strings.Contains(gotErr.Error(), wantClusterNACKErr) {
+		t.Fatalf("update received with error: %v, want %q", gotErr, wantClusterNACKErr)
+	}
+}
+
+// TestCDSWatch_PartialValid covers the case where a response from the
+// management server contains both valid and invalid resources and is expected
+// to be NACK'ed by the xdsclient. The test verifies that watchers corresponding
+// to the valid resource receive the update, while watchers corresponding to the
+// invalid resource receive an error.
+func (s) TestCDSWatch_PartialValid(t *testing.T) {
+	overrideFedEnvVar(t)
+	mgmtServer, nodeID, bootstrapContents, _, cleanup := e2e.SetupManagementServer(t, nil)
+	defer cleanup()
+
+	// Create an xDS client with the above bootstrap contents.
+	client, err := xdsclient.NewWithBootstrapContentsForTesting(bootstrapContents)
+	if err != nil {
+		t.Fatalf("Failed to create xDS client: %v", err)
+	}
+	defer client.Close()
+
+	// Register two watches for cluster resources. The first watch is expected
+	// to receive an error because the received resource is NACKed. The second
+	// watch is expected to get a good update.
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	badResourceName := cdsName
+	updateCh1 := testutils.NewChannel()
+	cdsCancel1 := client.WatchCluster(badResourceName, func(u xdsresource.ClusterUpdate, err error) {
+		updateCh1.Send(xdsresource.ClusterUpdateErrTuple{Update: u, Err: err})
+	})
+	defer cdsCancel1()
+	goodResourceName := cdsNameNewStyle
+	updateCh2 := testutils.NewChannel()
+	cdsCancel2 := client.WatchCluster(goodResourceName, func(u xdsresource.ClusterUpdate, err error) {
+		updateCh2.Send(xdsresource.ClusterUpdateErrTuple{Update: u, Err: err})
+	})
+	defer cdsCancel2()
+
+	// Configure the management server with two cluster resources. One of these
+	// is a bad resource causing the update to be NACKed.
+	resources := e2e.UpdateOptions{
+		NodeID: nodeID,
+		Clusters: []*v3clusterpb.Cluster{
+			badClusterResource(badResourceName, edsName, e2e.SecurityLevelNone),
+			e2e.DefaultCluster(goodResourceName, edsName, e2e.SecurityLevelNone)},
+		SkipValidation: true,
+	}
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+	}
+
+	// Verify that the expected error is propagated to the watcher which
+	// requested for the bad resource.
```

Comment:
s/ "watcher which requested for the bad resource" to "watching the bad resource" to keep consistent with comment below.